### PR TITLE
BUG: Groupby selection context not being properly reset

### LINKED ISSUE #28523: DataFrameGroupBy.first() affects group without assignment
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({'A' : ['foo', 'bar', 'foo', 'bar'], 'B' : [1, 2, 3, 4]})
grouped = df.groupby('A')
print(grouped.apply(lambda x: x))
grouped.first()
print(grouped.apply(lambda x: x))
```

Output:
```
     A  B
0  foo  1
1  bar  2
2  foo  3
3  bar  4

   B
0  1
1  2
2  3
3  4
```

#### Problem description

The invoking method, without re-assignment of result or specifying that it should work in place, should not affect the object upon which it is being called.

#### Expected Output
```
     A  B
0  foo  1
1  bar  2
2  foo  3
3  bar  4

     A  B
0  foo  1
1  bar  2
2  foo  3
3  bar  4
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-62-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.0
numpy            : 1.17.0
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.1
setuptools       : 40.6.3
Cython           : 0.27.3
pytest           : 3.0.7
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.2.6
html5lib         : None
pymysql          : None
psycopg2         : 2.7.3.2 (dt dec pq3 ext lo64)
jinja2           : 2.10
IPython          : 7.2.0
pandas_datareader: None
bs4              : 4.6.3
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.2.6
matplotlib       : 2.0.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : 0.11.0
pyarrow          : 0.13.0
pytables         : None
s3fs             : None
scipy            : 1.0.0
sqlalchemy       : 1.3.7
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by christopherzimmerman**
- [x] closes #28523 
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -240,6 +240,7 @@ Groupby/resample/rolling
 - Bug in :meth:`DataFrame.rolling` not allowing for rolling over datetimes when ``axis=1`` (:issue: `28192`)
 - Bug in :meth:`DataFrame.groupby` not offering selection by column name when ``axis=1`` (:issue:`27614`)
 - Bug in :meth:`DataFrameGroupby.agg` not able to use lambda function with named aggregation (:issue:`27519`)
+- Bug in :meth:`DataFrameGroupby` causing unexpected mutations of the groupby object (:issue:`28523`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]: I think the ref needs to be :class:\`pandas.core.groupby.DataFrameGroupBy\`. And maybe briefly explain the user-visible behavior: "Multiple operations on a DataFrameGroupBy object not giving the same results in certain cases".

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -433,6 +433,33 @@ def test_frame_groupby_columns(tsframe):
         assert len(v.columns) == 2
 
 
+@pytest.mark.parametrize(
+    "func, args",
+    [
+        ("sum", []),
```
Comment (TomAugspurger) [inline pandas/tests/groupby/test_groupby.py]: DO we need all these parametrezations? Or does just a few work? `sum` and `nth`?

Comment (christopherzimmerman) [reply pandas/tests/groupby/test_groupby.py]: Probably not.  It looks like one of [sum, prod, min, max, first, last] and then nth would suffice.

Comment (WillAyd) [reply pandas/tests/groupby/test_groupby.py]: We have a `reduction_func` fixture I think you should use here; would actually be more functions but also more comprehensive and consistent with some other tests

File: pandas/tests/groupby/test_groupby.py [hunk 3]
```diff
@@ -433,6 +433,33 @@ def test_frame_groupby_columns(tsframe):
...
+)
+def test_frame_groupby_avoids_mutate(func, args):
+    # GH28523
+    df = pd.DataFrame({"A": ["foo", "bar", "foo", "bar"], "B": [1, 2, 3, 4]})
+    grouped = df.groupby("A")
+
+    expected = grouped.apply(lambda x: x)
+
+    fn = getattr(grouped, func)
+    fn(*args)
```
Comment (WillAyd) [inline pandas/tests/groupby/test_groupby.py]: If `nth` is the only one that requires an argument I think OK to special case within the test

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks good just some minor comments. Did you see any other usages that weren't part of the context manager by chance?

**Comment by jbrockmendel** [comment]
Will this fix #26864?

File: pandas/tests/groupby/test_groupby.py (hunk 3 above)
Comment (christopherzimmerman) [reply pandas/tests/groupby/test_groupby.py]: @WillAyd the only other instance I looked at was inside of `_make_wrapper`, but it didn't seem related to this bug, and I haven't looked at it in depth.

**Comment by christopherzimmerman** [comment]
So there are two tests that will need to be changed regarding this fix:

[`test_grouper_creation_bug`](https://github.com/pandas-dev/pandas/blob/d4e2734ffb7a13fdf10f0ed8ad59521841e8ed07/pandas/tests/groupby/test_grouping.py#L164) and [`test_agg_timezone_round_trip`](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/groupby/aggregate/test_other.py#L457)

The second change is straightforward, the first might have identified another bug.

Is this behavior expected?

```
>>> df = DataFrame({"A": [0, 0, 1, 1, 2, 2], "B": [1, 2, 3, 4, 5, 6]})
>>> df.groupby('A').sum()
    B
A
0   3
1   7
2  11
>>> df.groupby('A').apply(lambda x: x.sum())
   A   B
A
0  0   3
1  2   7
2  4  11
```

The mutating state caused those two to be equal in the test, but now it fails, and I'm not sure if the second is intended behavior.

**Comment by WillAyd** [comment]
I think the second is a bug might have open issue already

**Comment by WillAyd** [comment]
#22546 #22542 #20420 might all be solved by this

**Comment by christopherzimmerman** [comment]
@WillAyd OK, I'll track down that bug and take care of that before I do anything else with this PR.  Also per your previous comment, making the test work with `reduction_func` removed the last instance of the improper set selection, so I think the context manager is now used everywhere.

**Comment by WillAyd** [comment]
Sounds good - nice work so far!

**Comment by WillAyd** [comment]
#17091 *might* be closed by this as well

**Comment by jreback** [comment]
@christopherzimmerman can you merge master, been some changed in groupby recently.

**Comment by jbrockmendel** [comment]
@christopherzimmerman can you rebase, this would be really nice to see fixed

**Comment by christopherzimmerman** [comment]
@jbrockmendel rebased and cleaned up what I was working on.  https://github.com/pandas-dev/pandas/issues/28549 still needs to be addressed before this is green.  I had a PR going for that issue and I'll revisit it.

**Comment by jbrockmendel** [comment]
a lot of work has been going on in these files; can you rebase again?  ill try to review in a timely manner so we dont have to go through another round

**Comment by jreback** [comment]
@christopherzimmerman would be a nice change, can you merge master

**Comment by christopherzimmerman** [comment]
I will merge, but #29131 will need to be merged first or else the CI will fail, since this requires some test updates from the other PR.

**Comment by WillAyd** [comment]
Let's come back to this after the other PR is merged then; (looks like more changes are needed there).

Closing to reduce PR queue for the time being - thanks as always!

**Pull request closed without merging**